File: src/main/java/org/c3lang/intellij/C3SyntaxHighlighterLexer.java
package org.c3lang.intellij;

import com.intellij.psi.tree.IElementType;
import org.c3lang.intellij.lexer.C3LexerAdapter;
import org.c3lang.intellij.psi.C3ElementType;
import org.c3lang.intellij.psi.C3Types;

public class C3SyntaxHighlighterLexer extends C3LexerAdapter
{
	private IElementType overrideTokenType = null;
	private IElementType prev = null;

	@Override
	public void advance()
	{
		super.advance();
		do
		{
			if (prev == CONST_IDENT_FAULT)
			{
				overrideTokenType = CONST_IDENT_FAULT_QUESTION;
				break;
			}
			if (getTokenType() == C3Types.CONST_IDENT
				&& getTokenEnd() < getBufferSequence().length()
				&& getBufferSequence().charAt(getTokenEnd()) == '~')
			{
				overrideTokenType = CONST_IDENT_FAULT;
				break;
			}
			overrideTokenType = null;
		} while (false);
		prev = getTokenType();
	}

	@Override
	public IElementType getTokenType()
	{
		return overrideTokenType != null ? overrideTokenType : super.getTokenType();
	}

	public static final C3ElementType CONST_IDENT_FAULT = new C3ElementType("CONST_IDENT_FAULT");
	public static final C3ElementType CONST_IDENT_FAULT_QUESTION = new C3ElementType("CONST_IDENT_FAULT_QUESTION");
}

File: src/main/java/org/c3lang/intellij/C3SyntaxHighlighterLexer.kt
package org.c3lang.intellij

import com.intellij.psi.tree.IElementType
import org.c3lang.intellij.lexer.C3LexerAdapter
import org.c3lang.intellij.psi.C3ElementType
import org.c3lang.intellij.psi.C3Types

class C3SyntaxHighlighterLexer : C3LexerAdapter() {

    private var overrideTokenType : IElementType? = null
    private var prev : IElementType? = null

    override fun advance() {
        super.advance()

        if (prev == CONST_IDENT_FAULT) {
            overrideTokenType = CONST_IDENT_FAULT_QUESTION
        } else if (tokenType == C3Types.CONST_IDENT && bufferSequence[tokenEnd] == '?' ) {
            overrideTokenType = CONST_IDENT_FAULT
        } else {
            overrideTokenType = null
        }

        prev = tokenType
    }

    override fun getTokenType(): IElementType? {
        return overrideTokenType ?: super.getTokenType()
    }

    companion object {
        @JvmField
        val CONST_IDENT_FAULT = C3ElementType("CONST_IDENT_FAULT")
        @JvmField
        val CONST_IDENT_FAULT_QUESTION = C3ElementType("CONST_IDENT_FAULT_QUESTION")
    }
}